src/format.rs

```diff
@@ -1,97 +1,247 @@
     use chrono::{Datelike, Timelike};
 
     let timestamp = chrono::Utc::now();
-    write!(
-        buf,
-        "{:004}-{:02}-{:02}T{:02}:{:02}:{:02}.{:06}Z ",
-        timestamp.year(),
-        timestamp.month(),
-        timestamp.day(),
-        timestamp.hour(),
-        timestamp.minute(),
-        timestamp.second(),
-        timestamp.nanosecond() / 1000,
-    )
-    .unwrap_or_else(|_| unreachable!());
+    let mut itoa = itoa::Buffer::new();
+
+    buf[4..8].copy_from_slice(itoa.format(timestamp.year()).as_bytes());
+    zero_pad2(&mut buf[9..11], itoa.format(timestamp.month()).as_bytes());
+    zero_pad2(&mut buf[12..14], itoa.format(timestamp.day()).as_bytes());
+    zero_pad2(&mut buf[15..17], itoa.format(timestamp.hour()).as_bytes());
+    zero_pad2(&mut buf[18..20], itoa.format(timestamp.minute()).as_bytes());
+    zero_pad2(&mut buf[21..23], itoa.format(timestamp.second()).as_bytes());
+    zero_pad6(
+        &mut buf[24..30],
+        itoa.format(timestamp.nanosecond() / 1000).as_bytes(),
+    );
 }
 
-/// Prints key values in ": key1=value1, key2=value2" format.
-///
-/// # Notes
-///
-/// Prints ": " itself, only when there is at least one key value pair.
-struct KeyValuePrinter<'a>(&'a dyn kv::Source);
-
-impl<'a> fmt::Display for KeyValuePrinter<'a> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        self.0
-            .visit(&mut KeyValueVisitor(true, f))
-            .map_err(|_| fmt::Error)
+#[inline]
+#[cfg(feature = "timestamp")]
+fn zero_pad2(buf: &mut [u8], v: &[u8]) {
+    debug_assert_eq!(buf.len(), 2);
+    if v.len() == 1 {
+        buf[0] = b'0';
+        buf[1] = v[0];
+    } else {
+        buf[0] = v[0];
+        buf[1] = v[1];
     }
 }
 
-struct KeyValueVisitor<'a, 'b>(bool, &'a mut fmt::Formatter<'b>);
+#[inline]
+#[cfg(feature = "timestamp")]
+fn zero_pad6(buf: &mut [u8], v: &[u8]) {
+    debug_assert_eq!(buf.len(), 6);
+    let start = 6 - v.len();
+    for i in 0..=start {
+        buf[i] = b'0';
+    }
+    buf[start..6].copy_from_slice(v);
+}
 
-impl<'a, 'b, 'kvs> kv::Visitor<'kvs> for KeyValueVisitor<'a, 'b> {
+/// Formats key value pairs in the following format: `key="value"`. For example:
+/// `user_name="Thomas" user_id=123 is_admin=true`
+struct KeyValueVisitor<'b>(&'b mut Vec<u8>);
+
+impl<'b, 'kvs> kv::Visitor<'kvs> for KeyValueVisitor<'b> {
     fn visit_pair(&mut self, key: kv::Key<'kvs>, value: kv::Value<'kvs>) -> Result<(), kv::Error> {
-        self.1
-            .write_str(if self.0 { ": " } else { ", " })
-            .and_then(|()| {
-                self.0 = false;
-                write!(self.1, "{}={}", key, value)
-            })
-            .map_err(Into::into)
+        self.0.push(b' ');
+        self.0.extend_from_slice(key.as_str().as_bytes());
+        self.0.push(b'=');
+        if let Some(value) = value.to_borrowed_str() {
+            self.0.push(b'\"');
+            self.0.extend_from_slice(value.as_bytes());
```

Comment:
Just wondering, is escaping a concern here if the value contains nested `"` characters?

Reply:
Yes it should be, for now I'm just sort of ignoring it. I don't really want to check the string, so I think I want to solve in the consumer of the logs. But I haven't made up my mind yet.